File: app/src/main/java/com/bubelov/coins/ui/widget/PlaceDetailsView.java
package com.bubelov.coins.ui.widget;

import android.content.Context;
import android.graphics.Paint;
import android.support.v7.widget.Toolbar;
import android.text.TextUtils;
import android.util.AttributeSet;
import android.view.*;
import android.widget.FrameLayout;
import android.widget.TextView;
import android.widget.ViewSwitcher;

import com.bubelov.coins.R;
import com.bubelov.coins.model.Currency;
import com.bubelov.coins.model.Place;
import com.bubelov.coins.util.Analytics;
import com.bubelov.coins.util.Utils;

import java.util.Iterator;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

/**
 * @author Igor Bubelov
 */

public class PlaceDetailsView extends FrameLayout {
    @BindView(R.id.header_switcher)
    ViewSwitcher headerSwitcher;

    @BindView(R.id.name)
    TextView name;

    @BindView(R.id.check_mark)
    View checkMark;

    @BindView(R.id.warning)
    View warning;

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.opened_claims)
    TextView openedClaims;

    @BindView(R.id.closed_claims)
    TextView closedClaims;

    @BindView(R.id.phone)
    TextView phone;

    @BindView(R.id.website)
    TextView website;

    @BindView(R.id.description)
    TextView description;

    @BindView(R.id.opening_hours)
    TextView openingHours;

    @BindView(R.id.accepted_currencies)
    TextView acceptedCurrencies;

    Place place;

    Listener listener;

    public PlaceDetailsView(Context context) {
        super(context);
        init();
    }

    public PlaceDetailsView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init();
    }

    public PlaceDetailsView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init();
    }

    public int getHeaderHeight() {
        return headerSwitcher.getHeight();
    }

    public void setFullScreen(boolean fullScreen) {
        if (fullScreen) {
            headerSwitcher.setDisplayedChild(1);
        } else {
            headerSwitcher.setDisplayedChild(0);
        }
    }

    public void setPlace(final Place place) {
        this.place = place;

        if (place.getOpenedClaims() > 0 && place.getClosedClaims() == 0) {
            checkMark.setVisibility(VISIBLE);
        } else {
            checkMark.setVisibility(GONE);
        }

        if (place.getClosedClaims() > 0) {
            warning.setVisibility(VISIBLE);
        } else {
            warning.setVisibility(GONE);
        }

        if (place.getOpenedClaims() > 0) {
            openedClaims.setVisibility(VISIBLE);
            openedClaims.setText(getResources().getQuantityString(R.plurals.confirmed_by_d_users, place.getOpenedClaims(), place.getOpenedClaims()));
        } else {
            openedClaims.setVisibility(GONE);
        }

        if (place.getClosedClaims() > 0) {
            closedClaims.setVisibility(VISIBLE);
            closedClaims.setText(getResources().getQuantityString(R.plurals.marked_as_closed_by_d_users, place.getClosedClaims(), place.getClosedClaims()));
        } else {
            closedClaims.setVisibility(GONE);
        }

        if (TextUtils.isEmpty(place.getName())) {
            name.setText(R.string.name_unknown);
            toolbar.setTitle(R.string.name_unknown);
        } else {
            name.setText(place.getName());
            toolbar.setTitle(place.getName());
        }

        if (TextUtils.isEmpty(place.getPhone())) {
            phone.setText(R.string.not_provided);
        } else {
            phone.setText(place.getPhone());
        }

        if (TextUtils.isEmpty(place.getWebsite())) {
            website.setText(R.string.not_provided);
            website.setTextColor(getResources().getColor(R.color.black));
            website.setPaintFlags(website.getPaintFlags() & (~ Paint.UNDERLINE_TEXT_FLAG));

            website.setOnClickListener(null);
        } else {
            website.setText(place.getWebsite());
            website.setTextColor(getResources().getColor(R.color.primary_dark));
            website.setPaintFlags(website.getPaintFlags() | Paint.UNDERLINE_TEXT_FLAG);
            website.setOnClickListener(view -> Utils.openUrl(PlaceDetailsView.this.getContext(), place.getWebsite()));
        }

        if (TextUtils.isEmpty(place.getDescription())) {
            description.setText(R.string.not_provided);
        } else {
            description.setText(place.getDescription());
        }

        if (TextUtils.isEmpty(place.getOpeningHours())) {
            openingHours.setText(R.string.not_provided);
        } else {
            openingHours.setText(place.getOpeningHours());
        }

        StringBuilder currenciesString = new StringBuilder();
        Iterator<Currency> currenciesIterator = place.getCurrencies().iterator();

        while (currenciesIterator.hasNext()) {
            currenciesString.append(currenciesIterator.next().getName());

            if (currenciesIterator.hasNext()) {
                currenciesString.append(", ");
            }
        }

        acceptedCurrencies.setText(currenciesString.length() == 0
                ? getContext().getString(R.string.not_provided)
                : currenciesString.toString());
    }

    public void setListener(Listener listener) {
        this.listener = listener;
    }

    private void init() {
        inflate(getContext(), R.layout.widget_place_details, this);
        ButterKnife.bind(this);

        toolbar.setNavigationOnClickListener(view -> {
            if (listener != null) {
                listener.onDismissed();
            }
        });

        toolbar.inflateMenu(R.menu.place_details);

        toolbar.setOnMenuItemClickListener(item -> {
            if (item.getItemId() == R.id.action_share) {
                Utils.share(getContext(), getResources().getString(R.string.share_place_message_title), getResources().getString(R.string.share_place_message_text, String.format("https://www.google.com/maps/@%s,%s,19z?hl=en", place.getLatitude(), place.getLongitude())));
                Analytics.logShareContentEvent(String.valueOf(place.getId()), place.getName(), "place");
                return true;
            }

            return false;
        });
    }

    public interface Listener {
        void onEditPlaceClick(Place place);

        void onDismissed();
    }

    @OnClick(R.id.edit)
    public void onEditClick() {
        if (listener != null) {
            listener.onEditPlaceClick(place);
        }
    }
}

File: app/src/main/java/com/bubelov/coins/ui/widget/PlaceDetailsView.kt
/*
 * This is free and unencumbered software released into the public domain.
 *
 * Anyone is free to copy, modify, publish, use, compile, sell, or
 * distribute this software, either in source code form or as a compiled
 * binary, for any purpose, commercial or non-commercial, and by any
 * means.
 *
 * In jurisdictions that recognize copyright laws, the author or authors
 * of this software dedicate any and all copyright interest in the
 * software to the public domain. We make this dedication for the benefit
 * of the public at large and to the detriment of our heirs and
 * successors. We intend this dedication to be an overt act of
 * relinquishment in perpetuity of all present and future rights to this
 * software under copyright law.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
 * EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
 * MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
 * IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
 * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
 * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
 * OTHER DEALINGS IN THE SOFTWARE.
 *
 * For more information, please refer to <https://unlicense.org>
 */

package com.bubelov.coins.ui.widget

import android.content.Context
import android.graphics.Paint
import android.support.v4.content.ContextCompat
import android.text.TextUtils
import android.util.AttributeSet
import android.view.*
import android.widget.FrameLayout
import android.widget.Toast

import com.bubelov.coins.R
import com.bubelov.coins.model.Place
import com.bubelov.coins.util.openUrl

import kotlinx.android.synthetic.main.widget_place_details.view.*
import android.content.Intent

class PlaceDetailsView(context: Context, attrs: AttributeSet) : FrameLayout(context, attrs) {
    internal lateinit var place: Place

    init {
        View.inflate(context, R.layout.widget_place_details, this)

        place_summary_toolbar.setNavigationOnClickListener { callback?.onDismissed() }
        place_summary_toolbar.inflateMenu(R.menu.place_details)

        place_summary_toolbar.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.action_share -> {
                    val subject = resources.getString(R.string.share_place_message_title)
                    val text = resources.getString(
                        R.string.share_place_message_text,
                        String.format(
                            "https://www.google.com/maps/@%s,%s,19z?hl=en",
                            place.latitude,
                            place.longitude
                        )
                    )

                    val intent = Intent(android.content.Intent.ACTION_SEND)
                    intent.type = "text/plain"
                    intent.putExtra(android.content.Intent.EXTRA_SUBJECT, subject)
                    intent.putExtra(android.content.Intent.EXTRA_TEXT, text)
                    context.startActivity(Intent.createChooser(intent, "Share"))

                    callback?.onShared(place)
                    true
                }
                else -> false
            }
        }

        edit.setOnClickListener { callback?.onEditPlaceClick(place) }
    }

    var fullScreen: Boolean = false
        set(value) {
            field = value

            if (value) {
                map_header_shadow.visibility = View.GONE
                map_header.visibility = View.GONE
                place_summary_toolbar.visibility = View.VISIBLE
            } else {
                map_header_shadow.visibility = View.VISIBLE
                map_header.visibility = View.VISIBLE
                place_summary_toolbar.visibility = View.GONE
            }
        }

    var callback: Callback? = null

    fun setPlace(place: Place) {
        this.place = place

        if (place.openedClaims > 0 && place.closedClaims == 0) {
            check_mark.visibility = VISIBLE
        } else {
            check_mark.visibility = GONE
        }

        if (place.closedClaims > 0) {
            warning.visibility = VISIBLE
        } else {
            warning.visibility = GONE
        }

        if (place.openedClaims > 0) {
            opened_claims.visibility = VISIBLE
            opened_claims.text = resources.getQuantityString(
                R.plurals.confirmed_by_d_users,
                place.openedClaims,
                place.openedClaims
            )
        } else {
            opened_claims.visibility = GONE
        }

        if (place.closedClaims > 0) {
            closed_claims.visibility = VISIBLE
            closed_claims.text = resources.getQuantityString(
                R.plurals.marked_as_closed_by_d_users,
                place.closedClaims,
                place.closedClaims
            )
        } else {
            closed_claims.visibility = GONE
        }

        if (TextUtils.isEmpty(place.name)) {
            name.setText(R.string.name_unknown)
            place_summary_toolbar.setTitle(R.string.name_unknown)
        } else {
            name.text = place.name
            place_summary_toolbar.title = place.name
        }

        if (TextUtils.isEmpty(place.phone)) {
            phone.setText(R.string.not_provided)
        } else {
            phone.text = place.phone
        }

        if (TextUtils.isEmpty(place.website)) {
            website.setText(R.string.not_provided)
            website.setTextColor(ContextCompat.getColor(context, R.color.black))
            website.paintFlags = website!!.paintFlags and Paint.UNDERLINE_TEXT_FLAG.inv()
            website.setOnClickListener(null)
        } else {
            website.text = place.website
            website.setTextColor(ContextCompat.getColor(context, R.color.primary_dark))
            website.paintFlags = website.paintFlags or Paint.UNDERLINE_TEXT_FLAG
            website.setOnClickListener {
                if (!context.openUrl(place.website)) {
                    Toast.makeText(context, "Can't open url: ${place.website}", Toast.LENGTH_SHORT)
                        .show()
                }
            }
        }

        if (TextUtils.isEmpty(place.description)) {
            description.setText(R.string.not_provided)
        } else {
            description.text = place.description
        }

        if (TextUtils.isEmpty(place.openingHours)) {
            opening_hours.setText(R.string.not_provided)
        } else {
            opening_hours.text = place.openingHours
        }
    }

    interface Callback {
        fun onEditPlaceClick(place: Place)
        fun onDismissed()
        fun onShared(place: Place)
    }
}